src/api/server.rkt

```diff
@@ -79,56 +79,98 @@
+     (place-channel-put manager (list 'timeline job-id b))
+     (log "Getting timeline for job: ~a.\n" job-id)
+     (place-channel-get a)]
+    [else
+     (define result (hash-ref single-threaded-cache job-id #f))
+     (if (false? result) (reverse (unbox (*timeline*))) result)]))
 
 ; Returns #f if there is no job returns the job-id if there is a completed job.
 (define (server-check-on job-id)
-  (define-values (a b) (place-channel))
-  (place-channel-put manager (list 'check job-id b))
-  (log "Checking on: ~a.\n" job-id)
-  (place-channel-get a))
+  (cond
+    [manager
+     (define-values (a b) (place-channel))
+     (place-channel-put manager (list 'check job-id b))
+     (log "Checking on: ~a.\n" job-id)
+     (place-channel-get a)]
+    [else (if (hash-has-key? single-threaded-cache job-id) job-id #f)]))
 
 (define (get-improve-table-data)
-  (define-values (a b) (place-channel))
-  (place-channel-put manager (list 'improve b))
-  (log "Getting improve results.\n")
-  (place-channel-get a))
+  (cond
+    [manager
+     (define-values (a b) (place-channel))
+     (place-channel-put manager (list 'improve b))
+     (log "Getting improve results.\n")
+     (place-channel-get a)]
+    [else
+     (for/list ([(job-id result) (in-hash single-threaded-cache)]
+                #:when (equal? (hash-ref result 'command) "improve"))
+       (get-table-data-from-hash result (make-path job-id)))]))
 
 (define (job-count)
-  (define-values (a b) (place-channel))
-  (place-channel-put manager (list 'count b))
-  (define count (place-channel-get a))
-  (log "Current job count: ~a.\n" count)
-  count)
+  (cond
+    [manager
+     (define-values (a b) (place-channel))
+     (place-channel-put manager (list 'count b))
+     (define count (place-channel-get a))
+     (log "Current job count: ~a.\n" count)
+     count]
+    [else (if job-running 1 0)])) ; We can only have one job running in this mode.
+
+(define job-running #f)
 
 ;; Starts a job for a given command object|
 (define (start-job command)
   (define job-id (compute-job-id command))
-  (place-channel-put manager (list 'start manager command job-id))
-  (log "Job ~a, Qed up for program: ~a\n" job-id (test-name (herbie-command-test command)))
+  (cond
+    [manager
+     (log "Job ~a, Qed up for program: ~a\n" job-id (test-name (herbie-command-test command)))
+     (place-channel-put manager (list 'start manager command job-id))]
+    [else
+     (log "Waiting for job ~a to finish.\n" job-id)
+     (define job-thread
+       (thread (λ ()
+                 (let loop ([seed #f])
+                   (match (thread-receive)
+                     [(list work job-id semaphore) (single-thread-herbie work job-id semaphore)])
+                   (loop seed)))))
+     (define sema (make-semaphore))
+     (set! job-running #t)
+     (thread-send job-thread (list command job-id sema))
+     (semaphore-wait sema)
+     (set! job-running #f)]) ;; Block for job to finish
```

Comment:
Why put this in a thread? Won't this thread keep running forever? Don't we only send it one job? Why have a thread at all?

Reply:
Also if there was no thread you wouldn't need a semaphor. And if you did that you wouldn't need the `single-thread-herbie` function which does very little on its own.